Source/WebCore/rendering/style/StyleCustomPropertyData.cpp

```diff
@@ -28,104 +28,149 @@
 {
+    ASSERT(!m_hasChildren);
+    ASSERT([&] {
+        auto* existing = get(name);
+        return !existing || !existing->equals(value);
+    }());
+
     auto addResult = m_ownValues.set(name, WTFMove(value));
-    if (addResult.isNewEntry && (!m_parentValues || !m_parentValues->m_ownValues.contains(name)))
-        ++m_ownValuesSizeExcludingOverriddenParentValues;
+
+    bool isNewProperty = addResult.isNewEntry && (!m_parentValues || !m_parentValues->get(name));
+    if (isNewProperty)
+        ++m_size;
 }
 
 bool StyleCustomPropertyData::operator==(const StyleCustomPropertyData& other) const
 {
-    if (size() != other.size())
+    if (m_size != other.m_size)
         return false;
 
-    for (auto& entry : m_ownValues) {
-        auto* otherValue = other.get(entry.key);
-        if (!otherValue || !entry.value->equals(*otherValue))
+    if (m_parentValues == other.m_parentValues) {
+        // This relies on the values in m_ownValues never being equal to those in m_parentValues.
+        if (m_ownValues.size() != other.m_ownValues.size())
             return false;
-    }
 
-    if (m_parentValues) {
-        ASSERT(!m_parentValues->m_parentValues);
-        for (auto& entry : m_parentValues->m_ownValues) {
-            if (m_ownValues.contains(entry.key))
-                continue;
-            auto* otherValue = other.get(entry.key);
+        for (auto& entry : m_ownValues) {
+            auto* otherValue = other.m_ownValues.get(entry.key);
             if (!otherValue || !entry.value->equals(*otherValue))
                 return false;
         }
+        return true;
     }
 
-    return true;
+    bool isEqual = true;
+    forEachInternal([&](auto& entry) {
+        auto* otherValue = other.get(entry.key);
+        if (!otherValue || !entry.value->equals(*otherValue)) {
+            isEqual = false;
+            return IterationStatus::Done;
+        }
+        return IterationStatus::Continue;
+    });
+
+    return isEqual;
 }
 
-void StyleCustomPropertyData::forEach(const Function<void(const KeyValuePair<AtomString, RefPtr<const CSSCustomPropertyValue>>&)>& callback) const
+template<typename Callback>
+void StyleCustomPropertyData::forEachInternal(Callback&& callback) const
 {
-    if (m_parentValues) {
-        ASSERT(!m_parentValues->m_parentValues);
-        for (auto& entry : m_parentValues->m_ownValues) {
-            if (!m_ownValues.contains(entry.key))
-                callback(entry);
+    Vector<const StyleCustomPropertyData*, maximumAncestorCount> descendants;
+
+    auto isOverridenByDescendants = [&](auto& key) {
+        for (auto* descendant : descendants) {
+            if (descendant->m_ownValues.contains(key))
+                return true;
+        }
+        return false;
+    };
+
+    auto* propertyData = this;
+    while (true) {
+        for (auto& entry : propertyData->m_ownValues) {
+            if (isOverridenByDescendants(entry.key))
```

Comment:
This is quadratic in the length of the ancestor chain, but since it's short, I guess it's OK in practice, and probably better than building up a `HashSet<AtomString>` of the properties we've seen so far.

Reply:
Right. With the limit in place the worst case is 4 hash lookup per value. I suspect that is faster than building a new map (which may be large).